kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/KieServerControllerImpl.java

```diff
@@ -62,10 +67,50 @@
 
     private static final Logger logger = LoggerFactory.getLogger(KieServerControllerImpl.class);
 
-    private KieServerTemplateStorage templateStorage = InMemoryKieServerTemplateStorage.getInstance();
+    private KieServerTemplateStorage templateStorage;
 
     private NotificationService notificationService = LoggingNotificationService.getInstance();
 
+    public KieServerControllerImpl () {
+        // retrieve the implementations
+        Map<String, PersistingServerTemplateStorageService> servicesFound = new HashMap<>();
+
+        ServiceLoader<PersistingServerTemplateStorageService> storageServices = ServiceLoader.load(PersistingServerTemplateStorageService.class);
+        Iterator<PersistingServerTemplateStorageService> storageServicesIterator = storageServices.iterator();
+        while (storageServicesIterator.hasNext()) {
+            PersistingServerTemplateStorageService storageService = storageServicesIterator.next();
+            logger.debug("Server template storage for kie server controller found {}", storageService.getTemplateStorage().toString());
+            servicesFound.put(storageService.getClass().getName(), storageService);
+        }
+
+        String storageServiceClassName = System.getProperty(KieServerConstants.KIE_SERVER_SERVER_TEMPLATE_STORAGE_SERVICE, "org.kie.server.controller.impl.service.FileBasedServerTemplateStorageService");
+        String []servicesClassNames = storageServiceClassName.split(",");
```

Comment:
Does it mean that user can specify several storage classes and one of those will be used when available on classparh?

Reply:
@sutaakar Correct. I am still not sure about the approach but yes.. It will search in order and It will be pick the first one.  I didn't put any QE guy just checking if the approach would be backward compatible. We have plenty of places were the service loader works not very determinstic way. I need to check which deployments deploys what jars to see if this make sense or just setting one of them is enough.

As I didn't test if open shift controller will be enough to sync among BC instances when servers deploy... I thinks so. Maybe you have more info I didn't have the time to test.

Any ideas are welcome

Reply:
There shouldn't be any significant problem with OpenShift controller storage as it is backed by ConfigMap which is shared with all Business central instances in openShift namespace.

Reply:
IMHO setting one storage service should be enough.

Reply:
removed that. by default file system is used.